### ISSUE TITLE
Bio.SubsMat.read_text_matrix unexpected result

### ISSUE THREAD

**Issue by mdehoon** [biopython/biopython#2202]
### Setup

I am reporting a problem with Biopython version, Python version, and operating
system as follows:

```python
CPython
Darwin-18.7.0-x86_64-i386-64bit
1.75.dev0
```

### Steps to reproduce

Download this Blast substitution matrix from NCBI:
ftp://ftp.ncbi.nlm.nih.gov/blast/matrices/NUC.4.4

Then load it with the `read_text_matrix` function in `Bio.SubsMat`:
```python
>>> handle = open("NUC.4.4")
>>> from Bio import SubsMat
>>> mat = SubsMat.read_text_matrix(handle)
>>> mat[('A','A')]
5.0
```

This is the correct value (compare to the file from NCBI). But for off-diagonal entries, all values were doubled. For example,
```python
>>> mat[('A','T')]
-8.0
```

while the file from NCBI shows -4 in this position.

**Comment by mdehoon** [comment]
@peterjc  @MarkusPiotrowski  This will affect users who use `read_text_matrix` to create a substitution matrix to use with the pairwise sequence aligners.

**Comment by peterjc** [comment]
This is related to the key order, and treatment as a full matrix vs a half matrix:

```
$ python
Python 3.7.3 | packaged by conda-forge | (default, Jul  1 2019, 14:38:56) 
[Clang 4.0.1 (tags/RELEASE_401/final)] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from Bio import SubsMat
>>> with open("NUC.4.4") as handle:
...     mat = SubsMat.read_text_matrix(handle)
... 
>>> print(mat['A', 'A'])
5.0
>>> print(mat['A', 'T'])
-8.0
>>> print(mat['T', 'A'])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: ('T', 'A')
```

It is the call to ``_full_to_half`` which combines ``mat['A', 'T'] == -4`` with ``mat['T', 'A'] == -4`` into a single entry of ``-8``, quoting the class ``__init__``:

```
        if build_later:
            self._init_zero()
        else:
            # Convert full to half
            self._full_to_half()
            self._correct_matrix()
```

If this was a mismatch in the parser and the class constructor expectations - treating it as a half matrix input (when it isn't) - it dates back a very long time. Was it deliberate?

**Comment by mdehoon** [comment]
Yes it was deliberate - see lines 40-43 in `Bio/SubsMat/__init__.py`. But it leads to surprising results, as in the example above.

I think that there are three fundamental problems with this module:
1) Unexpected behavior of half-matrices and full-matrices, as in the current issue;
2) The name of this module (`SubsMat`). The relevant class in this module is SeqMat. SubsMat refers to substitution matrix, and SubstitutionMatrix is a trivial subclass of SeqMat.
3) The fact that SeqMat is a subclass of dict, which makes it inconvenient for C modules (both for users and for developers). Subclassing from a (plain Python or numpy) array would work more elegantly.

Changing this module while keeping backward compatibility will be very difficult, so I think we should write a new module that over time will replace Bio.SubsMat.

**Comment by peterjc** [comment]
https://github.com/biopython/biopython/blob/biopython-174/Bio/SubsMat/__init__.py#L40

A new class seems a good idea.

**Comment by mdehoon** [comment]
What do you think would be a good place for a new class to live? I guess `seqmats` could be a reasonable name for a module, and `SeqMat` or `SequenceMatrix` as the class name. And since this module would be used from Bio.Align only, perhaps `seqmats` could be a submodule of Bio.Align?

**Comment by peterjc** [comment]
Are you thinking in terms of substitution matrices when you say sequence matrix? To me that sounds more like an array representation of a multiple sequence alignment. But a new top level module under ``Bio`` or a submodule under ``Bio.Align`` seems reasonable either way.

**Comment by mdehoon** [comment]
Fixed by PR #2297 by adding a submodule under `Bio.Align`; the issue in `Bio.SubsMat` remains.

**Comment by mdehoon** [comment]
Closing this, as `Bio.SubsMat` has been removed.
